app/server/lib/GristServerSocket.ts

```diff
@@ -89,40 +79,38 @@ export class GristServerSocketEIO extends GristServerSocket {
   public send(data: string, cb: (err?: Error) => void) {
     const msgNum = this._messageCounter++;
     this._messageCallbacks.set(msgNum, cb);
-    log.debug("sending msg %d: %s", msgNum, { toString:()=>data.slice(0, 100) });
     this._socket.send(data, {}, () => {
-      this._messageCallbacks.delete(msgNum);
-      log.debug("calling cb for msg %d", msgNum);
-      cb();
+      if (this._messageCallbacks.delete(msgNum)) {
+        cb?.();
```

Comment:
Good point. I changed `send` so that the callback is optional.

```ts
  public send(data: string, cb?: (err?: Error) => void) {
    const msgNum = this._messageCounter++;
    if (cb) {
      this._messageCallbacks.set(msgNum, cb);
    }
    this._socket.send(data, {}, () => {
      if (cb && this._messageCallbacks.delete(msgNum)) {
        cb();
      }
    });
  }
```